packages/babel-core/src/config/option-manager.js

```diff
@@ -64,29 +65,100 @@ const optionNames = new Set([
 ]);
 
 const ALLOWED_PLUGIN_KEYS = new Set([
-  "name",
+  "after",
+  "before",
+  "capabilities",
+  "inherits",
   "manipulateOptions",
-  "pre",
+  "name",
   "post",
+  "pre",
   "visitor",
-  "inherits",
 ]);
 
 export default function manageOptions(opts: {}): {
   options: Object,
-  passes: Array<Array<[ Plugin, ?{} ]>>,
+  plugins: Array<[ Plugin, ?{} ]>,
 }|null {
   return new OptionManager().init(opts);
 }
 
+function sortPlugins(plugins) {
+  const graph = new DagMap();
+  const capabilitiesToPluginIdMap = Object.create(null);
+
+  // It is theoretically possible that something arrives at this point without
+  // having a name or key property. Make sure that it has one.
+  const unnamedPluginPrefix = "unnamedPlugin";
+  plugins.forEach((pluginTuple, index) => {
+    const plugin = pluginTuple[0];
+    if (!plugin.key) { plugin.key = `${unnamedPluginPrefix}${index}`; }
+  });
+
+  // Calculate the capabilities that each plugin provides, map to a single name.
+  // Multiple plugins with the same capability are not allowed as that can
+  // become a cyclic dependency.
+  plugins.forEach((pluginTuple) => {
+    const plugin = pluginTuple[0];
+    if (plugin.capabilities) {
+      plugin.capabilities.forEach((capability) => {
+        const existingPluginName = capabilitiesToPluginIdMap[capability];
+        if (existingPluginName !== undefined) {
```

Comment:
prob safer to just say `if (existingPluginName)`, this way empty strings can also be caught. (although that shouldn't be allowed in the first place).